Examples of common Tabs API operations (#563)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/tabs/index.md

## Examples

> Opening the extension's onboarding page in new tab after the extension is installed for the first
> time. Must be called from a background context.

Reviewer: ```suggestion
A common pattern for extensions is to open an onboarding page in a new tab when the extension is installed. The following example shows how to do this. {% Aside %} Note that this code must be called from a background context, because tabs cannot use `chrome.tabs.create()`{% end Aside %}.
```

> Opening the extension's onboarding page in new tab after the extension is installed for the first
> time. Must be called from a background context.

Reviewer: ```suggestion
The following sections demonstrate several common use cases for the chrome.tabs API.

### Opening an extension page in a new tab

A common pattern for extensions is to open an onboarding page in a new tab when the extension is installed. The following example shows how to do this. {% Aside %} Note that this code must be called from a background context, because tabs cannot use `chrome.tabs.create()`{% end Aside %}.
```

```js
chrome.runtime.onInstalled.addListener((reason) => {
  if (reason === chrome.runtime.OnInstalledReason.INSTALL) {
    chrome.tabs.create({
      url: 'onboarding.html'
    });
  }
});
```

> Retrieve the user's currently focused tab. Requires Manifest V3. Cannot be used in a content

Reviewer 2: Why does it need V3? Purely because it uses `Promise`?

Author: Yep, exactly. TBH I'm a bit conflicted here. On one hand I want to provide snippets that are as broadly useful as possible, on the other I want to encourage modern language features. Perhaps it's enough if I just clarify this line to say "Requires Manifest V3 due to use of promises"?

Author: I should note that I also considered displaying both a promise & callback version of this snippet, but decided against it because it took up too much vertical space. Another use case in support of #90.

Reviewer 2: We talked about this; we want this to be a going forward document, so I think `Promise` only (with a note) makes sense.

Over the next year I presume we'll stop including the note as much.

Author: Yep, I think that makes sense.

> Retrieve the user's currently focused tab. Requires Manifest V3. Cannot be used in a content

Reviewer: ```suggestion
### Get the current tab

This example demonstrates how the background script can retrieve the currently focused tab. {% Aside %} Requires Manifest V3. Cannot be used in a content script. {% end Aside %}

{# Editor's note: what happened about converting this example to a promise with a note? #}
```

> script.

Reviewer: ```suggestion
```

```js
async function getCurrentTab() {
  let queryOptions = { active: true, currentWindow: true };
  let [tab] = await chrome.tabs.query(queryOptions);
  return tab;
}
```

> Content script asking a background script to navigate the current tab.

Reviewer: ```suggestion
### Request background script to navigate tab

This example shows how a content script could ask the background script to navigate the current tab.
{# Editor's note: I'd like to frame this as a use case; is there an obvious one? #}
```

```js
// Content script
function requestNavigation(url) {

>   chrome.runtime.sendMessage({ action: 'navigate', url });
> }
> 
> // Background script
> chrome.runtime.onMessage.addListener(messageListener);

Reviewer: ```suggestion
  chrome.runtime.sendMessage({ action: 'navigate', url });
}
<three back ticks>

<three back ticks>js
// Background script
chrome.runtime.onMessage.addListener(messageListener);
```

function messageListener(request, sender, _sendResponse) {
  if (request.action === 'navigate') {

>     // Sanitize URL in to guard against compromised content scripts

Reviewer 2: I mean.. we construct any URL we like from a string input. I don't think this counts as sanitizing.

Author: The idea was that an owned content script might try to pass in a non-URL payload in order to exploit either the extension's or Chrome's handling of this value. I thought that since `new URL()` will throw on invalid URLs, it could provide some basic sanitization. 

I did just did a little experimentation, though, and it looks like this constructor is more permissive than I originally thought. For example, I assumed new URL would throw on `javascript:alert(1)` or `data:text/html,<doctype html><script>alert(1)</script>` but neither cause the constructor to throw.

In general I want to encourage developers to think about how their inputs could be abused, especially messages from content scripts. If you have any thoughts on how to accomplish this I'd love to hear 'em.

Reviewer 3: There was an article by someone from Chromium on CORB topic noting that instead of sending full URLs in a message from a content script extensions should only pass an id or a path which will be expanded to a full URL in the background script. Like, this is the only way to make sure the background script doesn't act like an open proxy once the content script is compromised.

Reviewer 2: So I thought about this a bit more: _is_ there a security risk? This is just telling the user's browser to navigate somewhere, no different than if they typed it in the URL bar itself.

Like this is a pointless extension, I could just type in the URL. But I don't think this actually has a security problem. Am I missing something?

Reviewer 3: It could be a restricted chrome:// URL like `chrome://crash` or other dangerous ones listed in `chrome://chrome-urls` page.

Author: That's a good point, @tophf, thanks for calling that out.

@samthor, it's not as big an issue as, say, transmitting credit card info over HTTP, but it's still very much part of the extensions threat model. Content scripts run in the page's process, not the extensions. If a malicious actor is able to compromise the page's process, they also gains access to whatever capabilities extensions have on the page. That's why we limit what APIs are exposed to content scripts.

In that spirit, I want to encourage developers not to assume that data form content scripts can be trusted, and to think about ways to limit the severity of a potential compromise. To that end, I'm not as interested in the specifics of this example as I concise ways to demonstrate this line of thinking.

Author: Replaced this demo with a muting a given tab. I'd like to come back to this section to add a better message passing demo, but at the moment I no good use cases are leaping to mind. If anyone has thoughts, please add a suggestion to issue #665.

    let url = new URL(request.url);
    chrome.tabs.update(sender.tab.id, { url: url.toString() });
  }
}
```

### More samples

> ![Two tabs in a window](tabs.png)

Reviewer: I don't see the image on the preview...

Author: That image wasn't migrated in the transition to the new site. For the moment I'm just going to delete the image reference.

> More samples simple examples of manipulating tabs with the Tabs API can be found in the

Reviewer: ```suggestion
For more examples that demonstrate the Tabs API, see the
```

## Examples

The following sections demonstrate several common use cases for the chrome.tabs API.

### Opening an extension page in a new tab

A common pattern for extensions is to open an onboarding page in a new tab when the extension is
installed. The following example shows how to do this.

{% Aside %}

> Cannot use `chrome.tabs.create()` in content scripts.

Reviewer: Is a sentence with no subject. Consider any of:

- Content scripts cannot use `chrome.tabs.create()`
- `chrome.tabs.create()` is not available in content scripts.
- The `chrome.tabs.create()` method is not available in content scripts.

{% endAside %}.

```js
//// background.js

chrome.runtime.onInstalled.addListener((reason) => {
  if (reason === chrome.runtime.OnInstalledReason.INSTALL) {
    chrome.tabs.create({
      url: 'onboarding.html'
    });
  }
});
```

### Get the current tab

This example demonstrates how the background script can retrieve the currently focused tab.

{% Aside %}

> Requires Manifest V3 due to the use of Promises. Cannot use `tabs.query` in content

Reviewer: Are two sentences with no subject. (One less egregious.) Isn't "due to the use of promises" an implementation detail? Consider:

This example requires Manifest V3. Note also that the `tabs.query()` method isn't available in content scripts.

scripts.

> {% endAside %}

Author: > what happened about converting this example to a promise with a note?

@awfuchs, you included the above editor's note in a previous version, but I didn't quite follow the comment. The code uses Promises by awaiting the result of `chrome.tabs.query(queryOptions)`. Your suggested change also called out that the sample requires MV3. In a subsequent commit I clarified this line to say "Requires Manifest V3 due to the use of Promises."

Does this address your concerns?

Reviewer: Yes, I left new comments.

```js
//// background.js

async function getCurrentTab() {
  let queryOptions = { active: true, currentWindow: true };
  let [tab] = await chrome.tabs.query(queryOptions);
  return tab;
}
```

### Mute the specified tab

This example shows how an extension can toggle the muted state for a given tab.

{% Aside %}
Requires Manifest V3 due to the use of Promises. Cannot use `tabs.get` or `tabs.update`
in content scripts.
{% endAside %}

```js
//// background.js

function toggleMuteState(tabId) {
  chrome.tabs.get(tabId, async (tab) => {
    let muted = !tab.mutedInfo.muted;
    await chrome.tabs.update(tabId, { muted });
    console.log(`Tab ${tab.id} is ${ muted ? 'muted' : 'unmuted' }`);
  });
}

> ```

Author: > I'd like to frame this as a use case; is there an obvious one? 

FYI @awfuchs, I replaced the navigation request example with this new example and opened #665 to revisit a message passing example in the future.
